Add pass to rewrite pow2 div

gyulaz-htec opened the pull request:
This pass is needed for Llama2 fp16 where RMSNorm calc can go out of bounds outputting inf values.
The pass rewrites `x^2/n` to `(x/sqrt(n))^2`.

umangyadav commented on test/rewrite_pow2_div_test.cpp:
```diff
@@ -0,0 +1,199 @@
+
+#include <test.hpp>
+
+void run_pass(migraphx::module& m)
```
can this go in the same test file as simplify_algebra.cpp ?

umangyadav commented on test/simplify_algebra_test.cpp:
```diff
@@ -3711,4 +3711,32 @@ TEST_CASE(conv_concat)
     EXPECT(m1.sort() == m2.sort());
 }
 
+TEST_CASE(simplify_pow2_div)
```
add another test for integral type

pfultz2 commented:
This should go into a seperate pass. Maybe it could be called `rewrite_low_precision`?

pfultz2 commented on src/simplify_algebra.cpp:
```diff
@@ -1628,6 +1628,35 @@ struct find_split_transpose
+    auto matcher() const
+    {
+        return match::name("div")(
+            match::arg(0)(match::name("pow")(match::arg(0)(match::any().bind("x")),
```
This should also look for `mul(x, x)` as well.

pfultz2 commented on src/simplify_algebra.cpp:
```diff
@@ -1628,6 +1628,35 @@ struct find_split_transpose
+        auto x   = r.instructions["x"];
+
+        // sqrt(n) with integral numbers can cause huge accuracy loss due to float->int rounding
+        if(shape::is_integral(n->get_shape().type()))
```
We should also be checking for low precision. We probably shouldn't do this rewrite with fp32.

gyulaz-htec commented:
@pfultz2 I've updated the PR
* Moved the rewrite in the a pass (`rewrite_low_precision`).
* Updated the matcher to match `mul(x,x)`.
* Enabled only `fp16` for the rewrite, should I include `fp8` as well?
* Added test to cover `mul(x,x)` and some more to test that the new pass don't apply to other types or to `mul(x,y)`.

**I've updated the source branch, but github didn't sync my changes in the PR. If this doesn't change druing the weekend I can reopen it.**

pfultz2 commented on src/quantization.cpp:
```diff
@@ -57,6 +58,7 @@ void quantize_fp16(program& prog, const std::vector<std::string>& ins_names)
 {
     run_passes(prog,
                {normalize_ops{},
+                rewrite_low_precision{},
```
This pass should run after `rewrite_reduce`.

gyulaz-htec replied:
Thanks! That makes more sense.

pfultz2 commented on src/include/migraphx/matcher.hpp:
```diff
@@ -594,6 +594,14 @@ MIGRAPHX_PRED_MATCHER(same_input_shapes, instruction_ref ins)
+{
+    if(ins->inputs().empty())
+        return false;
+    auto s = ins->inputs().front();
```
Can you name this variable `input`? The `s` seems to imply a shape which it is not.

gyulaz-htec commented:
@pfultz2 I've made the proposed chanegs

lakhinderwalia commented on src/rewrite_low_precision.cpp:
```diff
@@ -0,0 +1,69 @@
+        auto ins = r.result;
+        auto n   = r.instructions["n"];
+        auto x   = r.instructions["x"];
+
```
If instead of:
  x^2/n --> (x/sqrt(n))^2, 
The following is applied, is there any loss of accuracy? 
  x^2/n --> (x/n) * n
Thanks.

gyulaz-htec replied:
Yes, that's actually works, the accuracy is the same with that. This solution saves us one instruction, also removes the `sqrt`. Thanks for the idea.

lakhinderwalia replied:
Thanks.

gyulaz-htec commented:
@pfultz2 @umangyadav @lakhinderwalia I've updated the PR with what @lakhinderwalia [proposed](https://github.com/ROCm/AMDMIGraphX/pull/2844#discussion_r1512126948) to use x^2/n --> (x/n) * x. The tests are passing with that change.

lakhinderwalia reviewed (approved):
Thanks.
